File: worker/src/worker/redis_client.py
"""Async Redis client wrapper."""
from __future__ import annotations

import asyncio
from typing import Any, Iterable

from redis import asyncio as redis_asyncio
from redis.exceptions import RedisError

from .config import get_settings
from .logger import get_logger
from .utils import with_retry

logger = get_logger(__name__)


class RedisClient:
    """Encapsulates Redis interactions with retry logic."""

    def __init__(self, url: str | None = None) -> None:
        settings = get_settings()
        self._url = url or settings.redis_url
        self._client = redis_asyncio.Redis.from_url(self._url, decode_responses=True)
        self._settings = settings
        self._lock = asyncio.Lock()

    @property
    def client(self) -> redis_asyncio.Redis:
        return self._client

    async def ensure_connection(self) -> None:
        await with_retry(
            self._client.ping,
            retries=self._settings.max_retries,
            base_delay=self._settings.retry_backoff_base,
            logger=logger,
            operation_name="redis_ping",
        )

    async def blpop(self, keys: list[str], timeout: int) -> tuple[str, str] | None:
        if not keys:
            await asyncio.sleep(timeout)
            return None

        async def _op() -> tuple[str, str] | None:
            return await self._client.blpop(*keys, timeout=timeout)

        try:
            return await _op()
        except RedisError as exc:
            logger.error("BLPOP failed", extra={"context_error": str(exc)})
            await asyncio.sleep(timeout)
            return None

    async def rpush(self, key: str, value: str) -> None:
        await with_retry(
            lambda: self._client.rpush(key, value),
            retries=self._settings.max_retries,
            base_delay=self._settings.retry_backoff_base,
            logger=logger,
            operation_name="redis_rpush",
        )

    async def set_heartbeat(self, worker_id: str, interval: int) -> None:
        ttl = max(interval * 2, interval + 5)
        try:
            await self._client.set(f"workers:heartbeat:{worker_id}", "alive", ex=ttl)
        except RedisError as exc:
            logger.warning("Heartbeat failed", extra={"context_error": str(exc)})

    async def record_failure(self, key: str, value: str) -> None:
        await with_retry(
            lambda: self._client.rpush(key, value),
            retries=self._settings.max_retries,
            base_delay=self._settings.retry_backoff_base,
            logger=logger,
            operation_name="redis_record_failure",
        )

    async def scan_brand_queues(self) -> list[str]:
        pattern = f"{self._settings.redis_queue_prefix}:*:chunks"
        cursor = 0
        results: list[str] = []
        try:
            while True:
                cursor, chunk = await self._client.scan(cursor=cursor, match=pattern, count=100)
                results.extend(chunk)
                if cursor == 0:
                    break
        except RedisError as exc:
            logger.error("Scanning brand queues failed", extra={"context_error": str(exc)})
        return sorted(set(results))

    async def get_spike_history(self, brand: str, cluster_id: int) -> list[int]:
        key = self._spike_key(brand, cluster_id)
        try:
            history = await self._client.lrange(key, 0, -1)
            return [int(item) for item in history]
        except RedisError as exc:
            logger.warning("Fetching spike history failed", extra={"context_error": str(exc)})
            return []

    async def append_spike_history(self, brand: str, cluster_id: int, value: int) -> None:
        key = self._spike_key(brand, cluster_id)
        try:
            async with self._lock:
                pipe = self._client.pipeline()
                pipe.lpush(key, value)
                pipe.ltrim(key, 0, 99)
                pipe.expire(key, self._settings.spike_history_ttl_sec)
                await pipe.execute()
        except RedisError as exc:
            logger.warning("Updating spike history failed", extra={"context_error": str(exc)})

    def _spike_key(self, brand: str, cluster_id: int) -> str:
        return f"{self._settings.redis_spike_prefix}:{brand}:{cluster_id}"

    async def close(self) -> None:
        await self._client.close()

File: worker-rs/src/redis_client.rs
use std::time::Duration;

use anyhow::Context;
use redis::{aio::ConnectionManager, AsyncCommands, Client, RedisResult};
use tokio::sync::Mutex;
use tokio::time::sleep;

#[derive(Clone)]
pub struct RedisClient {
    inner: std::sync::Arc<Mutex<ConnectionManager>>,
}

impl RedisClient {
    pub async fn new(url: &str) -> anyhow::Result<Self> {
        let client = Client::open(url.to_string()).context("Failed to create Redis client")?;
        let manager = client
            .get_tokio_connection_manager()
            .await
            .context("Failed to create Redis connection manager")?;
        Ok(Self {
            inner: std::sync::Arc::new(Mutex::new(manager)),
        })
    }

    pub async fn ensure_connection(&self) -> anyhow::Result<()> {
        let mut conn = self.inner.lock().await;
        redis::cmd("PING")
            .query_async::<_, ()>(&mut *conn)
            .await
            .context("Redis PING failed")
    }

    pub async fn blpop(&self, keys: &[String], timeout: Duration) -> anyhow::Result<Option<(String, String)>> {
        if keys.is_empty() {
            sleep(timeout).await;
            return Ok(None);
        }

        let mut conn = self.inner.lock().await;
        let timeout_secs = timeout.as_secs() as usize;
        let result: Option<(String, String)> = redis::cmd("BLPOP")
            .arg(keys)
            .arg(timeout_secs)
            .query_async(&mut *conn)
            .await
            .context("Redis BLPOP failed")?;
        Ok(result)
    }

    pub async fn rpush(&self, key: &str, value: &str) -> anyhow::Result<()> {
        let mut conn = self.inner.lock().await;
        redis::cmd("RPUSH")
            .arg(key)
            .arg(value)
            .query_async::<_, ()>(&mut *conn)
            .await
            .context("Redis RPUSH failed")
    }

    pub async fn record_failure(&self, key: &str, value: &str) -> anyhow::Result<()> {
        self.rpush(key, value).await
    }

    pub async fn scan_brand_queues(&self, prefix: &str) -> anyhow::Result<Vec<String>> {
        let pattern = format!("{prefix}:*:chunks");
        let mut cursor: u64 = 0;
        let mut results: Vec<String> = Vec::new();
        loop {
            let mut conn = self.inner.lock().await;
            let (next, chunk): (u64, Vec<String>) = redis::cmd("SCAN")
                .arg(cursor)
                .arg("MATCH")
                .arg(&pattern)
                .arg("COUNT")
                .arg(100)
                .query_async(&mut *conn)
                .await
                .context("Redis SCAN failed")?;
            drop(conn);
            results.extend(chunk);
            if next == 0 {
                break;
            }
            cursor = next;
        }
        results.sort();
        results.dedup();
        Ok(results)
    }

    pub async fn set_heartbeat(&self, worker_id: &str, interval: Duration) -> anyhow::Result<()> {
        let key = format!("workers:heartbeat:{worker_id}");
        let ttl = (interval.as_secs().saturating_mul(2).max(interval.as_secs() + 5)) as usize;
        let mut conn = self.inner.lock().await;
        redis::cmd("SET")
            .arg(&key)
            .arg("alive")
            .arg("EX")
            .arg(ttl)
            .query_async::<_, ()>(&mut *conn)
            .await
            .context("Redis heartbeat SET failed")
    }

    pub async fn get_spike_history(&self, prefix: &str, brand: &str, cluster_id: i32) -> anyhow::Result<Vec<i64>> {
        let key = format!("{prefix}:{brand}:{cluster_id}");
        let mut conn = self.inner.lock().await;
        let history: Vec<String> = redis::cmd("LRANGE")
            .arg(&key)
            .arg(0)
            .arg(-1)
            .query_async(&mut *conn)
            .await
            .context("Redis LRANGE failed for spike history")?;
        Ok(history
            .into_iter()
            .filter_map(|value| value.parse::<i64>().ok())
            .collect())
    }

    pub async fn append_spike_history(
        &self,
        prefix: &str,
        brand: &str,
        cluster_id: i32,
        value: i64,
        ttl: Duration,
    ) -> anyhow::Result<()> {
        let key = format!("{prefix}:{brand}:{cluster_id}");
        let mut conn = self.inner.lock().await;
        let mut pipe = redis::pipe();
        pipe.cmd("LPUSH").arg(&key).arg(value).ignore();
        pipe.cmd("LTRIM").arg(&key).arg(0).arg(99).ignore();
        pipe.cmd("EXPIRE")
            .arg(&key)
            .arg(ttl.as_secs() as usize)
            .ignore();
        pipe.query_async::<_, ()>(&mut *conn)
            .await
            .context("Redis pipeline failed for spike history")
    }
}
